TechnicalDocuments/Proposals/FlowAsStruct.md

```diff
@@ -0,0 +1,73 @@
+# Refactor `Flow` to be a Struct
+
+* Author: Martin Nygren
+* Review Manager: Anders Ha
+
+## Introduction
+
+This proposal is inspired by the [Pointfree](https://www.pointfree.co) episodes on Witness Oriented design, [33](https://www.pointfree.co/episodes/ep33-protocol-witnesses-part-1), [34](https://www.pointfree.co/episodes/ep34-protocol-witnesses-part-2), [35](https://www.pointfree.co/episodes/ep35-advanced-protocol-witnesses-part-1), [36](https://www.pointfree.co/episodes/ep36-advanced-protocol-witnesses-part-2) and [39](https://www.pointfree.co/episodes/ep39-witness-oriented-library-design).
+
+## Motivation
+
+Currently `Flow` is defined as a `Protocol` and implemented by a couple of helper classes (e.g. `EmptyFlow`, `NavigationFlow`, `ModalFlow`) that are made available as calculated properties (`Flow.empty`, `UINavigationController.navigationFlow`, `UIViewController.modalFlow`). This implementation already indicates that modelling the concept of a `Flow` as struct with function variables is closer to what we want then defining `Flow` as a protocol. We would have preferred to extend various `UIKit` classes if modelling `Flow` as a protocol is a natural choice.
+
+```
+extension UIViewController: Flow {
+  public func present(_ viewController: UIViewController,
+                      animated: Bool,
+                      completion: (() -> Void)?) {
+                self.present(viewController,
+                             animated: animated,
+                             completion: completion)
+         }
+
+  public func replace(with viewController: UIViewController,
+                      animated: Bool) {
+                present(viewController, animated: true)
+         }
+
+  public func dismiss(animated: Bool,
+                      completion: (() -> Void)?) {
+
+                if self.presentedViewController == nil {
+                    DispatchQueue.main.async { completion?() }
+                } else {
+                    self.dismiss(animated: animated, completion: completion)
+                }
+          }
+}
+```
+
+The above example would have several problems. What if we want to have a view controller that defines a flow for embedding one or more view controllers inside subviews? How sure are we that calling the completion block is something that we always want to do?
+
+Another reason to change `Flow` from a protocol to a struct is that it makes it easier to inject side effects without impacting any business logic. One example is a requirement to post screen view events to the analytics service. Another to save user journeys to a local database to support a personalised home screen.
```

Comment:
I think this is a legit motivation for the move, but the solution does not showcase any leverage of the composability over the current solution.

Also under the current framework, one could design mechanisms to inject side effects without any ergonomic impact:

```swift
struct SideEffectFlow: Flow {
    let wrapped: Flow
    let didStartPresenting: (() -> Void)?

    func present(_ viewController: UIViewController, animated: Bool, completion: (() -> Void)?) {
        wrapped.present(viewController, animated: animated, completion: completion)
        didStartPresenting?()
    }
}

extension Flow {
    public func on(didStartPresenting: (() -> Void)? = nil) {
        return SideEffectFlow(wrapped: self, didStartPresenting: didStartPresenting)
    }
}
```

Reply:
> Another reason to change `Flow` from a protocol to a struct is that it makes it easier to inject side effects without impacting any business logic.

@zzcgumn First, sorry for reviewing this late (was unfamiliar with this part of the codebase). Could you show specific example that shows how it’s improved?

Decorating the `Flow` the way Anders proposes seems to match your needs and keep existing infrastructure untouched.